sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py

```diff
@@ -318,13 +371,51 @@ def fit(self, X, y):
         del self._in_fit  # hard delete so we're sure it can't be used anymore
         return self
 
+    def _has_been_fitted(self):
+        return len(getattr(self, '_predictors', [])) > 0
+
+    def _clear_state(self):
+        """Clear the state of the gradient boosting model."""
+        if hasattr(self, 'train_score_'):
+            del self.train_score_
+        if hasattr(self, 'validation_score_'):
+            del self.validation_score_
+        if hasattr(self, '_rng'):
+            del self._rng
+
+    def _tolist_state(self):
+        """Convert all array attributes to lists."""
+        self.train_score_ = self.train_score_.tolist()
+        self.validation_score_ = self.validation_score_.tolist()
+
+    def _compute_subsample_indices(self, X_binned_train, rng,
```

Comment:
Let's call it `_compute_small_trainset_indices`

Also can you add a comment in the docstring:

> For efficiency, we need to subsample the training set to compute scores with scorers.
Also note that the returned indices are not expected to be the same between different calls to fit() in a warm start context, since the rng may have been consumed.